### PR TITLE
WT-13142 Sort alphabetically stat_data.py with s_all

### PR DESCRIPTION
**Pull request by zunyiliu**

### CODE REVIEW

File: dist/stat.py [hunk 1]
```diff
@@ -32,6 +34,39 @@ def check_unique_description(sorted_list):
...
+# For each Stat subclass check the names are sorted in alphabetical order
+##########################################
+def check_name_sorted(stat_list):
+    stat_dict = defaultdict(list)
+    for stat in stat_list:
+        stat_dict[type(stat)].append(stat)
+    for stat_type, stats in stat_dict.items():
+        sorted_stats = sorted(stats, key=lambda stat: remove_suffix_digits(stat.name))
+        for i in range(len(stats)):
+            if (sorted_stats[i].name != stats[i].name):
```
Comment (etienneptl) [inline dist/stat.py:55]:
```suggestion
            if sorted_stats[i].name != stats[i].name:
```

File: dist/stat.py [hunk 2]
```diff
@@ -32,6 +34,39 @@ def check_unique_description(sorted_list):
...
+##########################################
+# Remove trailing digits for a string.
+# Used as a lambda in sorting stats by name.
+##########################################
+def remove_suffix_digits(str):
+    return re.sub(r'\d+$', '', str)
+    
+##########################################
+# Check names are sorted:
+# For each Stat subclass check the names are sorted in alphabetical order
```
Comment (etienneptl) [inline dist/stat.py:46]: I think the second sentence is enough:
```suggestion
# For each Stat subclass check the names are sorted in alphabetical order.
```

File: dist/stat.py [hunk 3]
```diff
@@ -32,6 +34,39 @@ def check_unique_description(sorted_list):
             print("ERROR: repeated stat description in - '%s'" % (i.desc))
         temp = i.desc
 
+##########################################
+# Remove trailing digits for a string.
+# Used as a lambda in sorting stats by name.
```
Comment (etienneptl) [inline dist/stat.py:39]: I don't think we need to say how it's used currently as it may change overtime.
```suggestion
```

File: dist/stat.py [hunk 4]
```diff
@@ -32,6 +34,39 @@ def check_unique_description(sorted_list):
...
+# Check names are sorted:
+# For each Stat subclass check the names are sorted in alphabetical order
+##########################################
+def check_name_sorted(stat_list):
+    stat_dict = defaultdict(list)
+    for stat in stat_list:
+        stat_dict[type(stat)].append(stat)
+    for stat_type, stats in stat_dict.items():
+        sorted_stats = sorted(stats, key=lambda stat: remove_suffix_digits(stat.name))
+        for i in range(len(stats)):
```
Comment (etienneptl) [inline dist/stat.py:54]: I think we could use [zip()](https://docs.python.org/3/library/functions.html#zip) to traverse both lists at the same time.

File: dist/stat.py [hunk 5]
```diff
@@ -32,6 +34,39 @@ def check_unique_description(sorted_list):
...
+
+all_stat_list = [conn_stats, dsrc_stats, conn_dsrc_stats, join_stats, session_stats]
+for stat_list in all_stat_list:
+    check_name_sorted(stat_list)
+
+conn_stats.sort(key=attrgetter('desc'))
+dsrc_stats.sort(key=attrgetter('desc'))
+conn_dsrc_stats.sort(key=attrgetter('desc'))
+join_stats.sort(key=attrgetter('desc'))
+session_stats.sort(key=attrgetter('desc'))
```
Comment (etienneptl) [inline dist/stat.py:68]: I suggest to sort them alphabetically:
```suggestion
conn_stats.sort(key=attrgetter('desc'))
conn_dsrc_stats.sort(key=attrgetter('desc'))
dsrc_stats.sort(key=attrgetter('desc'))
join_stats.sort(key=attrgetter('desc'))
session_stats.sort(key=attrgetter('desc'))
```

File: dist/stat_data.py [hunk 6]
```diff
@@ -93,6 +93,14 @@ class CursorStat(Stat):
...
     def __init__(self, name, desc, flags=''):
         Stat.__init__(self, name, CursorStat.prefix, desc, flags)
+class CursorSweepStat(Stat):
+    prefix = 'cursor'
+    def __init__(self, name, desc, flags=''):
+        Stat.__init__(self, name, CursorStat.prefix, desc, flags)
+class CursorErrorStat(Stat):
+    prefix = 'cursor'
+    def __init__(self, name, desc, flags=''):
+        Stat.__init__(self, name, CursorStat.prefix, desc, flags)
```
Comment (etienneptl) [inline dist/stat_data.py:103]: Let's sort those too:
```suggestion
class CursorErrorStat(Stat):
    prefix = 'cursor'
    def __init__(self, name, desc, flags=''):
        Stat.__init__(self, name, CursorStat.prefix, desc, flags)
class CursorSweepStat(Stat):
    prefix = 'cursor'
    def __init__(self, name, desc, flags=''):
        Stat.__init__(self, name, CursorStat.prefix, desc, flags)
```

**Review by etienneptl — COMMENTED** [review]
Nice job! I left a few comments/suggestions.

File: dist/stat.py [hunk 7]
```diff
@@ -32,6 +34,39 @@ def check_unique_description(sorted_list):
...
+
 check_unique_description(conn_stats)
 check_unique_description(dsrc_stats)
 check_unique_description(session_stats)
```
Comment (ajmorton) [inline dist/stat.py:72]: We mention in the ticket that we'll remove the existing `sorted` calls but I see they're still present ~10 lines below this comment. Have we decided to leave them in?

File: dist/stat.py (hunk 3 above)
Comment (ajmorton) [reply dist/stat.py:39]: Can we also have some context for why we strip strings here? 
Something like "Sometimes we have statistics like hist_100 and hist_50 where we want hist_50 to come first, but alphabetically hist_100 comes first. We rely on the developer to order these statistics correctly."

**Review by ajmorton — COMMENTED** [review]
Looking good! I'd like to confirm this comment before I review further.

File: dist/stat.py [hunk 8]
```diff
@@ -32,11 +34,44 @@ def check_unique_description(sorted_list):
...
         temp = i.desc
 
+##########################################
+# Remove trailing digits for a string.
+##########################################
+def remove_suffix_digits(str):
+    return re.sub(r'\d+$', '', str)
+    
+##########################################
+# For each Stat subclass check the names are sorted in alphabetical order.
```
Comment (kommiharibabu) [inline dist/stat.py:44]: I don't see a need for `S` in stat. It can be a lowe case also.

File: dist/stat.py [hunk 9]
```diff
@@ -32,11 +34,44 @@ def check_unique_description(sorted_list):
...
+        # There're stats like hist_100 and hist_50 where alphabetically hist_100 comes first. 
+        # Thus we remove suffix digits and rely on the developer to order these stats correctly.
+        sorted_stats = sorted(stats, key=lambda stat: remove_suffix_digits(stat.name))
+        for sorted_stat, stat in zip(sorted_stats, stats):
+            if sorted_stat.name != stat.name:
+                print(f"ERROR: {stat_type.__name__} not sorted alphabetically by name, expected " \
+                      f"'{sorted_stat.name}' but found '{stat.name}'")
+                return
+
+all_stat_list = [conn_stats, dsrc_stats, conn_dsrc_stats, join_stats, session_stats]
```
Comment (kommiharibabu) [inline dist/stat.py:60]: How about `conn_dsrc_stats` before `dsrc_stats` inside the array?

**Review by kommiharibabu — APPROVED** [review]
LGTM with two minor comments

File: dist/stat.py (hunk 8 above)
Comment (etienneptl) [inline dist/stat.py:44]:
```suggestion
# For each stat subclass check the names are sorted in alphabetical order.
```

File: dist/stat.py [hunk 10]
```diff
@@ -32,11 +34,44 @@ def check_unique_description(sorted_list):
...
+##########################################
+# For each Stat subclass check the names are sorted in alphabetical order.
+##########################################
+def check_name_sorted(stat_list):
+    stat_dict = defaultdict(list)
+    for stat in stat_list:
+        stat_dict[type(stat)].append(stat)
+    for stat_type, stats in stat_dict.items():
+        # There're stats like hist_100 and hist_50 where alphabetically hist_100 comes first. 
+        # Thus we remove suffix digits and rely on the developer to order these stats correctly.
```
Comment (etienneptl) [inline dist/stat.py:52]: I'd suggest the following comment to stay generic:
```suggestion
        # In alphabetical order, stat_name_100 comes before stat_name_50. For this reason, remove any numerical suffix before sorting the stats. Print an error if the stats are not sorted correctly.
```

File: dist/stat.py (hunk 9 above)
Comment (etienneptl) [inline dist/stat.py:60]:
```suggestion
all_stat_list = [conn_stat, conn_dsrc_stats, dsrc_stats, join_stats, session_stats]
```

File: dist/stat.py [hunk 11]
```diff
@@ -32,11 +34,44 @@ def check_unique_description(sorted_list):
...
+                print(f"ERROR: {stat_type.__name__} not sorted alphabetically by name, expected " \
+                      f"'{sorted_stat.name}' but found '{stat.name}'")
+                return
+
+all_stat_list = [conn_stats, dsrc_stats, conn_dsrc_stats, join_stats, session_stats]
+for stat_list in all_stat_list:
+    check_name_sorted(stat_list)
+
+conn_stats.sort(key=attrgetter('desc'))
+conn_dsrc_stats.sort(key=attrgetter('desc'))
```
Comment (etienneptl) [inline dist/stat.py:65]:
```suggestion
conn_dsrc_stats.sort(key=attrgetter('desc'))
conn_stats.sort(key=attrgetter('desc'))
```

**Review by etienneptl — COMMENTED** [review]
A few more comments, good to see your changes have found unsorted stats!

File: dist/stat.py (hunk 7 above)
Comment (ajmorton) [reply dist/stat.py]: Discussed this in Slack. The new sorting logic in this PR is purely for sorting the contents of `stat_data.py`, while these `sorted` control the order of stats produced by WiredTiger at runtime. I'm happy to leave these sorted calls as they are.

**Review by etienneptl — APPROVED** [review]
LGTM! Thank you for addressing all the comments.

**Review by etienneptl — APPROVED** [review]

**Review by ajmorton — APPROVED** [review]
LGTM 👍

**Comment by zunyiliu** [comment]
evergreen retry

**Pull request merged**